src/index.js

```diff
@@ -140,17 +141,39 @@ export default function nodeResolve ( options = {} ) {
 		}
 		const pkgRoot = dirname( pkgPath );
 
+		const packageInfo = {
+			// copy as we are about to munge the `main` field of `pkg`.
+			packageJson: Object.assign({}, pkg),
```

Comment:
This would definitely be gold-plating, but an approach that would avoid the `Object.assign` would be to change the code to not override `main` but instead e.g. take it from the `packageInfo`. I admit I did not look at the usages of `main`, though. What do you think? It would be cleaner IMO.

Reply:
overriding `main` is done mainly (tee hee) for controlling browserify/resolve's behavior. It doesn't handle e.g. different `main` fields like `module`, so support for that added by munging the package json before `resolve` sees it. Since the munging is somewhat complex, doing it up front and caching the result seems good to me.